functions/visualizations.py

```diff
@@ -169,6 +161,116 @@ def _clustergrammerify(
+                metadata_df.pop(c)
+                continue
+
+        if "protocol_identifier" in c or c in ["participant_id"]:
+            # always duplicates because of structure of samples.csv and participants.csv
+            metadata_df.pop(c)
+            continue
+
+        if "(1=Yes,0=No)" in c:
+            # these are boolean! let's treat them that way
+            metadata_df[c] = metadata_df[c].astype(bool)
+
+    # rename the columns to pretty things
+    columns = []
+    for cat in metadata_df.columns:
+
+        if cat.startswith("arbitrary_trial_specific_clinical_annotations."):
+            # for 10021 participants.csv:
+            ## remove the prefix
+            ## remove any parentheses
+
+            cat = cat[46:]
+            if '(' in cat and ')' in cat and cat.index(')') > cat.index('('):
+                cat = cat.split('(',1)[0] + cat.rsplit(')',1)[1]
+        else:
+            # otherwise
+            ## break up underscores
+            ## title case
+            ## drop 'CIDC' / 'CIMAC' anywhere
+            ## drop trailing 'Name'
+            cat = (
+                cat.replace("_", " ")
+                .title()
+                .replace("Cidc", "")
+                .replace("Cimac", "")
+            )
+            if cat.endswith("Name") and not cat == "Name":
+                cat = cat[:-4]
+
+        # strip so it's pretty!
+        columns.append(cat.strip())
+    metadata_df.columns = columns
+
+    # don't cut down more if 3 or less survive
+    # these are probably the ones we want kept above
+    if metadata_df.shape[1] > 3:
+    # Remove the second of any pair of columns where they lead to the
+    # same groupings, combining into first column if values are different
+        m = 0
+        while m < metadata_df.shape[1]:
+            n = 1
+            while n+m < metadata_df.shape[1]:
+                # for every unique pairing of columns
+
+                c, d = metadata_df.columns[[m,m+n]] # get the two column names
+                if len(metadata_df[c].unique()) == len(metadata_df[d].unique()):
+                    # to be the same, they have the same number of uniques
+                    
+                    # get the groupings in a comparable form
+                    group_c = [tuple(metadata_df.index[metadata_df[c] == u]) for u in metadata_df[c].unique()]
+                    group_d = [tuple(metadata_df.index[metadata_df[d] == u]) for u in metadata_df[d].unique()]
+                    if group_c == group_d: # they are guaranteed to be in the same order if they're the same
+                        if (metadata_df[c] == metadata_df[d]).all():
+                            # if exactly the same, no need to keep both
+                            metadata_df.pop(d)
+                        else:
+                            # then combine the data
+                            metadata_df[c] = metadata_df[c].astype(str) + " / " + metadata_df.pop(d).astype(str)
```

Comment:
not sure what this should do, but it does this sometimes `Treatment / arbitrary_trial_specific_clinical_annotations.Randomized Arm`